### PR TITLE
Create single script to upload and validate pcluster s3 configs

### PR DESCRIPTION
**Pull request by demartinofra**
This script replaces instance-whitelist.py and upload-instance-slot-map.py and adds some extra checks, validations and dryrun mode.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: util/update-pcluster-configs.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    def validate_document(old_doc, new_doc):
+        try:
+            import dictdiffer
+
+            logging.info("Found following diffs: %s", list(dictdiffer.diff(old_doc, new_doc)))
+        except ImportError:
+            logging.warning("Install dictdiffer if you want to display diffs: pip install dictdiffer")
+
+        if S3JsonDocumentManager._assert_document_is_included(old_doc, new_doc):
+            raise Exception("The new configuration file does not include the old entries.")
```
Comment (ddeidda) [inline util/update-pcluster-configs.py]: Could we have a quick chat about this? I am not sure I got the intention

Comment (demartinofra) [reply util/update-pcluster-configs.py]: There is no intention :D forgot to change it when I made _assert_document_is_included raise on failures. Fixing

Comment (ddeidda) [reply util/update-pcluster-configs.py]: :+1:

File: util/update-pcluster-configs.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+                with open(pricing_file) as data_file:
+                    self.__pricing_file_cache = json.load(data_file)
+            else:
+                self.__pricing_file_cache = self._download_pricing_file(region)
+        return self.__pricing_file_cache
+
+    @staticmethod
+    def _download_pricing_file(region):
+        logging.info("Downloading pricing file...This might take a while.")
+        url_prefix = f"https://pricing.{region}.amazonaws.com"
```
Comment (lukeseawalker) [inline util/update-pcluster-configs.py]: attention to cn

File: util/update-pcluster-configs.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
+import collections
```
Comment (lukeseawalker) [inline util/update-pcluster-configs.py]: shebang and copyright missing

File: util/update-pcluster-configs.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+            s3 = boto3.resource("s3", region_name=self._region, **self._credentials)
+            instances_file_content = s3.Object(s3_bucket, document_s3_path).get()["Body"].read()
+            return json.loads(instances_file_content)
+        except Exception as e:
+            logging.error(
+                "Failed when downloading file %s from bucket %s in region %s with error %s",
+                document_s3_path,
+                s3_bucket,
+                self._region,
+                e,
```
Comment (lukeseawalker) [inline util/update-pcluster-configs.py]: maybe the error message is enough? Also in the next Exception

File: util/update-pcluster-configs.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+
+class ConfigGenerator(ABC):
+    @abstractmethod
+    def generate(self, args, region, credentials):
+        pass
+
+
+class InstancesConfigGenerator(ConfigGenerator):
+    SCHEMA = {
```
Comment (lukeseawalker) [inline util/update-pcluster-configs.py]: NICE!

File: util/update-pcluster-configs.py (hunk 2 above)
Comment (demartinofra) [reply util/update-pcluster-configs.py]: fixed

File: util/update-pcluster-configs.py (hunk 4 above)
Comment (demartinofra) [reply util/update-pcluster-configs.py]: not sure what the exception message is going to be so I'd rather leave and extra clear logging message

**Pull request merged**